src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.cs

```diff
@@ -69,39 +52,152 @@ protected virtual SyntaxNode GetNodeToAnalyze(SyntaxNode root, TextSpan span)
                 {
-                    actions.Add(GetCodeAction(state, operationKind: OperationKind.RenameType));
+                    actions.Add(GetCodeAction(state, fileName: state.DocumentName, operationKind: OperationKind.RenameType));
                 }
             }
 
             return actions.ToImmutableArray();
         }
 
-        private CodeAction GetCodeAction(
-            State state,
-            OperationKind operationKind)
+        private CodeAction GetCodeAction(State state, string fileName, OperationKind operationKind) =>
+            new MoveTypeCodeAction((TService)this, state, operationKind, fileName);
+
+        private bool IsNestedType(TTypeDeclarationSyntax typeNode) =>
+            typeNode.Parent is TTypeDeclarationSyntax;
+
+        /// <summary>
+        /// checks if there is a single top level type declaration in a document
+        /// </summary>
+        private bool MultipleTopLevelTypeDeclarationInSourceDocument(SyntaxNode root) =>
+            TopLevelTypeDeclarations(root).Count() > 1;
+
+        private static IEnumerable<TTypeDeclarationSyntax> TopLevelTypeDeclarations(SyntaxNode root) =>
+            root.DescendantNodes(n => (n is TCompilationUnitSyntax || n is TNamespaceDeclarationSyntax))
+                .OfType<TTypeDeclarationSyntax>();
+
+        private bool AnyTopLevelTypeMatchesDocumentName(State state, CancellationToken cancellationToken)
         {
-            return new MoveTypeCodeAction((TService)this, state, operationKind);
+            var root = state.SemanticDocument.Root;
+            var semanticModel = state.SemanticDocument.SemanticModel;
+
+            return TopLevelTypeDeclarations(root).Any(
+                typeDeclaration =>
+                {
+                    var typeName = semanticModel.GetDeclaredSymbol(typeDeclaration, cancellationToken).Name;
+                    return TypeMatchesDocumentName(typeDeclaration, typeName, state.DocumentName, semanticModel, cancellationToken);
+                });
         }
+
+        /// <summary>
+        /// checks if type name matches its parent document name, per style rules.
+        /// </summary>
+        /// <remarks>
+        /// Note: For a nested type, a matching document name could be just the type name or a
+        /// dotted qualified name of its type hierarchy.
+        /// </remarks>
+        protected bool TypeMatchesDocumentName(
+            TTypeDeclarationSyntax typeNode,
+            string typeName,
+            string documentName,
+            SemanticModel semanticModel,
+            CancellationToken cancellationToken)
+        {
+            // trim extensions, if present.
+            documentName = Path.GetFileNameWithoutExtension(documentName);
+
+            // If it is not a nested type, the comparison is straight forward.
+            if (!IsNestedType(typeNode))
+            {
+                return typeName.Equals(documentName, StringComparison.CurrentCulture);
```

Comment:
both codepaths do this check.  So you can just lift it out and do that check first.  If it fails, then do the dotted check.